include/deal.II/lac/read_write_vector.h

```diff
@@ -0,0 +1,709 @@
+     * data around.
+     */
+    template <typename ForwardIterator, typename OutputIterator>
+    void extract_subvector_to (ForwardIterator          indices_begin,
+                               const ForwardIterator    indices_end,
+                               OutputIterator           values_begin) const;
+
+    /**
+     * Read access to the data field specified by @p local_index. The largest
+     * range of the IndexSet is the first one.
+     *
+     * Performance: Direct array access (fast).
+     */
+    Number local_element (const size_type local_index) const;
+
+    /**
+     * Read and write access to the data field specified by @p local_index. The
+     * largest range of the IndexSet is the first one.
+     *
+     * Performance: Direct array access (fast).
+     */
+    Number &local_element (const size_type local_index);
+    //@}
+
+
+    /**
+     * @name 3: Modification of vectors
+     */
+    //@{
+
+    /**
+     * A collective add operation: This function adds a whole set of values
+     * stored in @p values to the vector components specified by @p indices.
+     */
+    template <typename OtherNumber>
+    void add (const std::vector<size_type>   &indices,
+              const std::vector<OtherNumber>  &values);
+
+    /**
+     * This is a second collective add operation. As a difference, this
+     * function takes a ReadWriteVector of values.
+     */
+    template <typename OtherNumber>
+    void add (const std::vector<size_type>        &indices,
+              const ReadWriteVector<OtherNumber> &values);
+
+    /**
+     * Take an address where <tt>n_elements</tt> are stored contiguously and
+     * add them into the vector. Handles all cases which are not covered by
+     * the other two <tt>add()</tt> functions above.
+     */
+    template <typename OtherNumber>
+    void add (const size_type    n_elements,
+              const size_type   *indices,
+              const OtherNumber  *values);
+
+    /**
+     * Prints the vector to the output stream @p out.
+     */
+    void print (std::ostream       &out,
+                const unsigned int  precision  = 3,
+                const bool          scientific = true) const;
+
+    /**
+     * Returns the memory consumption of this class in bytes.
+     */
+    std::size_t memory_consumption () const;
+    //@}
+
+  protected:
+
+    /**
+     * Return the local position of @p global_index.
+     */
+    unsigned int
+    global_to_local (const types::global_dof_index global_index) const
+    {
+      // the following will throw an exception if the global_index is not
+      // in the remaining_elements
+      return static_cast<unsigned int>(stored_elements.index_within_set (global_index));
+    }
+
+    /**
+     * A helper function that is used to resize the val array.
+     */
+    void resize_val (const size_type new_allocated_size);
+
+    /**
+     * The size that is currently allocated in the val array.
+     */
+    size_type allocated_size;
```

Comment:
Let's get rid of this member. In the old days, we used it to re-allocate memory only if someone resized a vector to something *larger* but not do any re-allocation to something smaller. In reality, resizing vectors is a rather rare thing, and resizing it to something smaller is even rarer. Let's just assume that `val` simply points to something that has size `stored_elements.size()`.